src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+                let node_radius = &shadow_size + 1.px() * dim.y();
+                let node_shape  = Rect(
+                    (&shadow_size*2.0 + 2.px() * dim.x(),&node_radius*2.0)).corners_radius(node_radius);
+                let node_shape  = node_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
                 let tx       = - 1.px() * pos.x();
                 let ty       = - 1.px() * pos.y();
-                let n_shape  = n_shape.translate((tx,ty));
+                let node_shape  = node_shape.translate((tx,ty));
 
-                let shape = shape * n_shape;
-                let shape = shape.fill(color::Rgba::from($($color)*));
+                let shape = shape.intersection(node_shape);
 
-                shape.into()
+                let split_shape = FocusedEdge::new(
+                shape,&focus_split_center.px(),&focus_split_angle.into());
+                let shape       = split_shape.fill($color, $highlight_color);
+
+                let hover_width = width + HOVER_EXTENSION.px() * 2.0;
+                let hover_area  = corner_base_shape(&radius,&hover_width,&angle,&start_angle);
+                let hover_area  = hover_area.fill(HOVER_COLOR);
+                (hover_area + shape).into()
+            }
+        }
+
+        impl EdgeShape for component::ShapeView<Shape> {
+            fn set_focus_split_center_local(&self, center:Vector2<f32>) {
+                self.shape.focus_split_center.set(center);
+            }
+
+            fn set_focus_split_angle(&self, angle:f32) {
+                 self.shape.focus_split_angle.set(angle);
+            }
+
+            fn sprite(&self) -> &Sprite{
+                &self.shape.sprite
+            }
+
+            fn events(&self) -> &ShapeViewEvents{
+                &self.events
+            }
+
+            fn normal_local(&self, point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+                point_rotation(point)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                let radius = self.shape.radius.get();
+                let center = Vector2::zero();
+                let point_to_center = point.xy() - center;
+
+                let closest_point = center + point_to_center / point_to_center.magnitude() * radius;
+                let vector_angle  = -nalgebra::Rotation2::rotation_between(&Vector2::new(0.0, 1.0),&closest_point).angle();
+                let start_angle   =  self.shape.start_angle.get();
+                let end_angle     =  start_angle + self.shape.angle.get();
+                let upper_bound   = start_angle.max(end_angle);
+                let lower_bound   = start_angle.min(end_angle);
+
+                let correct_quadrant = lower_bound < vector_angle && upper_bound > vector_angle;
+                if correct_quadrant {
+                     Some(Vector2::new(closest_point.x, closest_point.y))
+                } else {
+                    None
+                }
```

Comment:
in such cases use always `as_some` from boolinator exported in proelude

Reply:
What is the advantage of that over the explicit `Some(...)`?
Besides I am always getting a not-implemented error for `as_some`. Need to investigate why.

Reply:
Because the code is much nicer!

```rust
if correct_quadrant {
                     Some(Vector2::new(closest_point.x, closest_point.y))
                } else {
                    None
                }
```
vs
```rust
correct_quadrant.as_some(Vector2::new(closest_point.x,closest_point.y))
```